Give back some power to the end users (#288)
cosmiconfig/cosmiconfig: README.md

.{NAME}rc
.{NAME}rc.json
.{NAME}rc.yaml
.{NAME}rc.yml
.{NAME}rc.js
.{NAME}rc.cjs
.config/{NAME}rc
.config/{NAME}rc.json
.config/{NAME}rc.yaml
.config/{NAME}rc.yml
.config/{NAME}rc.js
.config/{NAME}rc.cjs
{NAME}.config.js
{NAME}.config.cjs
```

The contents of these files are defined by the tool.
For example, you can configure prettier to enforce semicolons at the end of the line
using a file named `.config/prettierrc.yml`:

```yaml
semi: true
```

Additionally, you have the option to put a property named after the tool in your `package.json` file,
with the contents of that property being the same as the file contents. To use the same example as above:

```json
{
  "name": "your-project",
  "dependencies": {},
  "prettier": {
    "semi": true
  }
}
```

This has the advantage that you can put the configuration of all tools
(at least the ones that use cosmiconfig) in one file.

You can also add a `cosmiconfig` key within your `package.json` file or create one of the following files
to configure `cosmiconfig` itself:

```
.config.json
.config.yaml
.config.yml
.config.js
.config.cjs
```

The following property is currently actively supported in these places:

```yaml
cosmiconfig:

>   # overrides where configuration files are being searched to enforce a common naming convention and format
>   # (which are defined by yourself)

Reviewer: ```suggestion
  # overrides where configuration files are being searched to enforce a custom naming convention and format
```

- Options.
- Asynchronous by default (though can be run synchronously).

## Usage for end users

When configuring a tool, you can use multiple file formats and put these in multiple places.

Usually, a tool would mention this in its own README file,
but by default, these are the following places, where the `moduleName` variable represents the name of the tool:

```js
[
  'package.json',
  `.${moduleName}rc`,
  `.${moduleName}rc.json`,
  `.${moduleName}rc.yaml`,
  `.${moduleName}rc.yml`,
  `.${moduleName}rc.js`,
  `.${moduleName}rc.cjs`,
  `.config/${moduleName}rc`,
  `.config/${moduleName}rc.json`,
  `.config/${moduleName}rc.yaml`,
  `.config/${moduleName}rc.yml`,
  `.config/${moduleName}rc.js`,
  `.config/${moduleName}rc.cjs`,
  `${moduleName}.config.js`,
  `${moduleName}.config.cjs`,

> ]

Reviewer: Perhaps this would be clearer if in the same format as the `searchPlaces` config:
```
package.json
.{name}rc
.{name}rc.json
...
```

Author: `searchPlaces` actually uses the same format as this, but I made it a bit clearer anyway since this is end user documentation.

```

The contents of these files are defined by the tool.
For example, you can configure prettier to enforce semicolons at the end of the line
using a file named `.config/prettierrc.yml`:

```yaml
semi: true
```

Additionally, you have the option to put a property named after the tool in your `package.json` file,
with the contents of that property being the same as the file contents. To use the same example as above:

```json
{
  "name": "your-project",
  "dependencies": {},
  "prettier": {
    "semi": true
  }
}
```

This has the advantage that you can put the configuration of all tools
(at least the ones that use cosmiconfig) in one file.

You can also add a `cosmiconfig` key within your `package.json` file or create one of the following files
to configure `cosmiconfig` itself:

```
.config.json
.config.yaml
.config.yml
.config.js
.config.cjs
```

The following properties are currently actively supported in these places:

```yaml
cosmiconfig:
  # overrides where configuration files are being searched, this might be a minor performance optimization
  # if all your configuration files use the same naming convention and format
  searchPlaces:

>     - .config/{name}.yml

Reviewer: This isn't one of the formats listed above; was it supposed to be `.config/{name}rc.yml`, or is this a separate search location?

Author: This is supposed to show that you can do whatever you want here.

Reviewer: Ah, in that case it might be worthwhile to consider replacing the blurb about "minor performance optimization". I don't think end users will really care about a ms-level startup improvement.

      
  # enables configuration resolution in this same file, similarly to `package.json`, see below

>   searchInThisFile: true

Reviewer: Any reason for this to not always be on?

Author: Good point, I think I had a reason, but I forgot. Will re-check.

Author: I see now - the reason why this exists is because it will conflict with the `ignoreEmptySearchPlaces` configuration that tool authors can set. If that is set, and searching in the meta config file is enabled (i.e. `searchInThisFile`), the explorer will not look anywhere else and return an empty configuration. I will add a note for this.

Author: Actually, I guess I'll fix that instead.

Reviewer: Ah, so that means that if you add
```yaml
# .config.yml
cosmiconfig:
  searchInThisFile: true
```
It may cause some subset of tools using cosmiconfig to ignore previously-working config files? That seems non-ideal. I'd only expect behavior like that on something like
```yaml

# .config.yml
cosmiconfig:
  searchInThisFile: true
  
prettier:
```

Author: I fixed this issue and removed the `searchInThisFile` flag.
